core-tests/shared/src/test/scala/zio/ZManagedSpec.scala

```diff
@@ -0,0 +1,760 @@
+            ZManagedSpecUtil
+              .testReservePar(2, res => ZManaged.mergeAllParN(2)(List.fill(4)(res))(0) { case (a, _) => a })
+          },
+          testM("Uses at most n fibers for acquisition") {
+            ZManagedSpecUtil
+              .testAcquirePar(2, res => ZManaged.mergeAllParN(2)(List.fill(4)(res))(0) { case (a, _) => a })
+          },
+          testM("Runs finalizers") {
+            ZManagedSpecUtil
+              .testFinalizersPar(4, res => ZManaged.mergeAllParN(2)(List.fill(4)(res))(0) { case (a, _) => a })
+          },
+          testM("All finalizers run even when finalizers have defects") {
+            for {
+              releases <- Ref.make[Int](0)
+              _ <- ZManaged
+                    .mergeAllParN(2)(
+                      List(
+                        ZManaged.finalizer(ZIO.dieMessage("Boom")),
+                        ZManaged.finalizer(releases.update(_ + 1)),
+                        ZManaged.finalizer(ZIO.dieMessage("Boom")),
+                        ZManaged.finalizer(releases.update(_ + 1)),
+                        ZManaged.finalizer(ZIO.dieMessage("Boom")),
+                        ZManaged.finalizer(releases.update(_ + 1))
+                      )
+                    )(())((_, _) => ())
+                    .use_(ZIO.unit)
+                    .run
+              count <- releases.get
+            } yield assert(count, equalTo(3))
+          }
+        ),
+        suite("onExit")(
+          testM("Calls the cleanup") {
+            for {
+              finalizersRef <- Ref.make[List[String]](Nil)
+              resultRef     <- Ref.make[Option[Exit[Nothing, String]]](None)
+              _ <- ZManaged
+                    .make(UIO.succeed("42"))(_ => finalizersRef.update("First" :: _))
+                    .onExit(e => finalizersRef.update("Second" :: _) *> resultRef.set(Some(e)))
+                    .use_(ZIO.unit)
+              finalizers <- finalizersRef.get
+              result     <- resultRef.get
+            } yield assert(finalizers, equalTo(List("Second", "First"))) && assert(
+              result,
+              isSome(equalTo(Exit.succeed("42")))
+            )
+          }
+        ),
+        suite("onExitFirst")(
+          testM("Calls the cleanup") {
+            for {
+              finalizersRef <- Ref.make[List[String]](Nil)
+              resultRef     <- Ref.make[Option[Exit[Nothing, String]]](None)
+              _ <- ZManaged
+                    .make(UIO.succeed("42"))(_ => finalizersRef.update("First" :: _))
+                    .onExitFirst(e => finalizersRef.update("Second" :: _) *> resultRef.set(Some(e)))
+                    .use_(ZIO.unit)
+              finalizers <- finalizersRef.get
+              result     <- resultRef.get
+            } yield assert(finalizers, equalTo(List("First", "Second"))) && assert(
+              result,
+              isSome(equalTo(Exit.succeed("42")))
```

Comment:
You can use `succeeds`, `fails` and `isInterrupted` to apply predicates on `Exit`.